### ISSUE TITLE
Specifying GPU rank on multi-GPU systems

### ISSUE THREAD

**Issue by afiaka87** [replicate/cog#583]
Laion.ai recently has been donated access to a few 8xA100 pods, of which I'm able to use. 

Tinkering with using `cog` for inference on these is tricky because it is a _shared_ instance. This means, whenever I log in - someone may be using some amount of the GPU's. As such, it's useful to be able to specify the device globally, rather than edit each line referencing the device rank in the framework of your choosing.

The framework for me is essentially always pytorch, which respects the `CUDA_VISIBLE_DEVICES` parameter allowing you to specify the "world" of GPU's for reference. `CUDA_VISIBLE_DEVICES=6,7` will show up to pytorch as `torch.device("cuda:0")` and `torch.device("cuda:1")`, respectively. This is more useful for training multi-gpu, but it also allows you to override the rank 0 GPU (i.e. `torch.device("cuda")`), which is commonly used throughout the code to move memory around in inference scripts.

In an ideal world, this would be specifiable as a parameter to the prediction script. In practice, hardcoded calls `.to("cuda")` and the equivalent `.cuda()` are common. People tend to develop on single-GPU or on a multi-GPU where they have access to the rank 0 GPU, after all.

### tl;dr:
Running cog containers on a multi-gpu environment is tricky assuming you don't want to change the repo's code.

### Solutions?

We could expose the user's environment to the docker container This could pose strange security issues?

We could allow users to manually set environment variables.

Perhaps the `cog` command could respect a `cog.env` file if it exists?
Alternatively, a `-e` flag might work here (if it isn't already used). For instance:

`cog predict -i text="my prompt" -i batch_size=4 -e "CUDA_VISIBLE_DEVICES=4"`.

What do you think?

**Comment by allenhung1025** [comment]
Hi @afiaka87,
When I use `-e` flag, it doesn't work. 
May you explain more about how you specify GPU rank on multi-gpu systems?

**Comment by allenhung1025** [comment]
Hi @afiaka87,
When I was trying to run this [model](https://replicate.com/allenhung1025/looptest) on multi-gpu systems, I got some error `Starting Docker image cog-looptest-base and running setup()...
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].`
<img width="1434" alt="Screen Shot 2022-05-23 at 4 21 05 PM" src="https://user-images.githubusercontent.com/22190213/169775908-c1d7bce4-5be0-4f75-b10d-2f5f921b3124.png">

**Comment by Shivam010** [comment]
Experiencing the same issue: `docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].` but with a different model.

@allenhung1025 were you able to resolve it ?

**Comment by markphelps** [comment]
This issue has been inactive for over three years. The original request was about passing environment variables (e.g., `CUDA_VISIBLE_DEVICES`) to specify GPU rank in `cog predict` on shared multi-GPU systems, while subsequent comments diverged into a separate nvidia-docker runtime configuration issue (`could not select device driver`). If you still need this capability on a current version of Cog, please open a new issue with details about your environment and the specific command you're running.
